exla/lib/exla/defn/stream.ex

```diff
@@ -0,0 +1,56 @@
+defmodule EXLA.Defn.Stream do
+  @moduledoc false
+  @derive {Inspect, only: [:executable, :keep_on_device, :send, :recv]}
+  @enforce_keys [:send, :send_shape, :recv, :recv_shape, :done, :executable, :keep_on_device]
+  defstruct [:send, :send_shape, :recv, :recv_shape, :done, :executable, :keep_on_device]
+
+  defimpl Nx.Stream do
+    def send(%{executable: executable, send: send, send_shape: send_shape}, data) do
+      unless Nx.compatible?(send, data) do
+        raise ArgumentError, """
+        Nx stream expected a tensor of type, shape, and names on send:
+
+        #{inspect(send)}
+
+        But got tensor:
+
+        #{inspect(data)}
+        """
+      end
+
+      pred = EXLA.Shape.make_shape({:pred, 8}, {})
+      :ok = EXLA.Client.to_infeed(executable.client, executable.device_id, <<0::8-native>>, pred)
```

Comment:
shouldn't this hard matches qualify for a `send!` function name instead of `send`? Where is the line drawn?

Reply:
Most of the time a bang function exists as an alternative to a non bang variant. In this case, they always succeed because when they fail it is usually due to an argument error or something intractable.

Reply:
I see! This makes everything much more clear!